File: javascript/0789-minimum-distance-between-two-bst-nodes.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var minDiffInBST = function (root) {
    let [prev, res] = [null, Number.MAX_VALUE];

    function dfs(node) {
        if (node) {
            dfs(node.left);

            if (prev) {
                res = Math.min(res, node.val - prev.val);
            }
            prev = node;

            dfs(node.right);
        }
    }

    dfs(root);

    return res;
};

File: go/0789-minimum-distance-between-two-bst-nodes.go
package main

import "math"

func main() {

}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func minDiffInBST(root *TreeNode) int {
	var prev *TreeNode
	res := math.MaxInt

	var dfs func(*TreeNode)

	dfs = func(node *TreeNode) {
		if node != nil {
			dfs(node.Left)

			if prev != nil {
				res = min(res, node.Val-prev.Val)
			}
			prev = node

			dfs(node.Right)
		}
	}

	dfs(root)
	return res
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
